File: 1. Data_Structure/1. Stack_Queue/StackQueue_1.java

import java.util.ArrayList;
import java.util.List;
// 다리를 지나는 트럭.
class Solution {
    public static int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;

        List<Integer> bridge_list = new ArrayList<> ();
        List<Integer> truck_list = new ArrayList<> ();
       
        for (int i = 0; i < bridge_length; i++) {
            bridge_list.add(i, 0);
        }

        for (int j = 0; j < truck_weights.length; j++) {
            truck_list.add(j, truck_weights[j]);
        }
        int tmp_sum = 0;

        while (bridge_list.size() != 0) {
            answer += 1;
            int tmp_remove = bridge_list.remove(0);
            tmp_sum -= tmp_remove;

            if (truck_list.size() != 0) {

                if (tmp_sum + truck_list.get(0) <= weight) {
                    tmp_sum += truck_list.get(0);
                    int tmp_add = truck_list.remove(0);
                    bridge_list.add(tmp_add);
                } else {
                    bridge_list.add(0);

                }

            }

        }
        
        return answer;
    }

    public static void main(String[] args) {
        int[] abc = { 7, 4, 5, 6 };
        System.out.println(solution(2, 10, abc));
        
    }




}

File: 1. Data_Structure_DW/1. Stack_Queue/StackQueue_1.py
## 다리를 지나는 트럭
# 2, 10, [7,4,5,6]
# sum()
def solution(bridge_length, weight, truck_weights):
    answer = 0
    tmp_truck = truck_weights[::-1]
    tmp_bridge = [0] * bridge_length
    sum_total = sum(tmp_bridge) ## 0
    while tmp_bridge:
        answer +=1
        sum_tmp = tmp_bridge.pop(0)
        sum_total -= sum_tmp ## 0
        if tmp_truck:
            if (sum_total + tmp_truck[-1]) <= weight:
                sum_total += tmp_truck[-1]
                tmp_bridge.append(tmp_truck.pop())
            else :
                tmp_bridge.append(0)
            
    return answer


print(solution(2, 10, [7,4,5,6]))
#print(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]))
print(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]))

# 서진원
def solution(bridge_length, weight, truck_weights):
    answer = 0 
    bridge_list = [0] * bridge_length
    while bridge_list :
        answer += 1 
        bridge_list.pop(0) 
        if truck_weights :
            if sum(bridge_list) + truck_weights[0] <= weight :
                bridge_list.append(truck_weights.pop(0))
            else :
                bridge_list.append(0)

    return answer 